src/basic/lock-util.c

```diff
@@ -186,3 +187,18 @@ int lock_generic(int fd, LockType type, int operation) {
                 assert_not_reached();
         }
 }
+
+int lock_generic_new_fd(int primary_fd, LockType type, int operation) {
+        _cleanup_close_ int lock_fd = -EBADF;
+        int r;
+
+        lock_fd = fd_reopen(ASSERT_FD(primary_fd), O_RDONLY|O_CLOEXEC|O_NONBLOCK|O_NOCTTY);
+        if (lock_fd < 0)
+                return lock_fd;
+
+        r = lock_generic(lock_fd, type, operation);
+        if (r < 0)
+                return r;
+
+        return TAKE_FD(lock_fd);
+}
```

Comment:
I think this is already covered by `xopenat_lock()` so no need to add a new helper (just need to drop `assert(path)` from `xopenat_lock()`).

```
xopenat_lock(fd, NULL, O_RDONLY|O_CLOEXEC|O_NONBLOCK|O_NOCTTY, 0, 0, LOCK_BSD, LOCK_EX);
```